### ISSUE TITLE
Manually generate signed url for react-native background upload issue

### ISSUE THREAD

**Issue by bneigher** [aws-amplify/amplify-js#5339]
Context: react-native, amplify-js 2.3.0
iOS and Android

** Which Category is your question related to? **
Auth, Storage

** What AWS Services are you utilizing? **
Cognito, Pinpoint, Storage.

Question:
I'm trying to leverage more powerful native features to avoid the pitfalls of mobile usage / e.g. uploading while in the background.  I'm following a guide here:
https://github.com/Vydia/react-native-background-upload/issues/55

```javascript

/*
staticMedia[mediaObject] = {
   uri: "file:///test/path/image.jpg,
   ...
}
*/
import Upload from 'react-native-background-upload';
import { Auth } from 'amplify-js';
import { Signer } from '@amplify-js/core';

const fileInfo = await Upload.getFileInfo(
  staticMedia[mediaObject].uri,
);

const credentials = await Auth.currentCredentials();
const cognitoUser = await Auth.currentAuthenticatedUser();

const url = `https://${Config.AWS_MEDIA_BUCKET}.s3.${Config.AWS_REGION}.amazonaws.com/private/${cognitoUser.id}/testDir/${fileInfo.name}`;

const signedRequest = Signer.sign(
  {
    url,
    headers: {
      'content-type': fileInfo.mimeType,
      'content-length': `${fileInfo.size}`,
    },
  },
  {
    access_key: credentials.data.Credentials.AccessKeyId,
    secret_key: credentials.data.Credentials.SecretKey,
    session_token: credentials.data.Credentials.SessionToken,
  },
  {
    service: 's3',
    region: Config.AWS_REGION,
  },
);

const options = {
   type: 'multipart',
   method: 'POST',
   headers: signedRequest.headers,
   url: signedRequest.url,
   field: 'file',
   path: staticMedia[mediaObject].uri,
};
const uploadId = await Upload.startUpload(options);
```

This works thus far in gathering all of the requirements (that I can see from the source / sdk).. But I'm seeing an error regarding one of the headers not being present.  Namely 
- :warning: Missing required header for this request: x-amz-content-sha256

I can see when Amplify Storage makes the request, it adds the `X-Amz-Content-Sha256` header, but the signer used in amplify core doesn't generate this header.  I tried also using `Signer.signUrl` but it complains nontheless about a signature mismatch.

When adding manually `X-Amz-Content-Sha256`: `UNSIGNED-PAYLOAD`, we have a signature mismatch (probably because it IS a signed payload?)

Also tried the alternative `STREAMING-AWS4-HMAC-SHA256-PAYLOAD` - still invalid signature :(

**Comment by mauerbac** [comment]
Hi @bneigher - a few thoughts after speaking with the team:

- It looks like you are on v2, so you can try to use the underlying aws-sdk [here](https://github.com/aws/aws-sdk-js/blob/master/lib/services/s3.js#L1035).
- If you want to migrate to v3, you could use [this](migrate to V3, they can use https://github.com/aws/aws-sdk-js-v3/tree/master/packages/s3-request-presigner).

Today, Amplify only supports presigned urls for the GET operation. Let me know if the above solutions would work.

**Comment by cameronrr** [comment]
The x-amz-content-sha256 header is required for all AWS Signature Version 4 requests. It provides a hash of the request payload. **If there is no payload, you must provide the hash of an empty string.**

Source: https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html

In your first argument to sign() your headers should include the following (the hash of an empty string):
```
headers: {
    // ... other headers
    'x-amz-content-sha256': 
        // this is the SHA256 of an empty string ("")
        'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855',
}
```

**Comment by ahansson89** [comment]
@bneigher, where you ever able to get this working? I am looking to do the same

**Comment by bneigher** [comment]
@ahansson89 I have moved off of amplify because of the issues I faced with react native.  I ultimately didn't trust the libraries ability to handle my token refreshing / session rotation as it was constantly failing me (I'm using auth0 as a federated identity) and ran into too many issues.  Painful experience.


But my code looks like this now:

```javascript
import Upload from 'react-native-background-upload';

const options = {
  type: 'raw',
  method: 'PUT',
  url: signedUrl, // generated with my API
  path: filePath, // local path to image
  customUploadId: `${fileInfo.name}`, // identifier
  // android
  maxRetries: 5,
  notification: {
    enabled: false,
    autoclear: true,
    autoClear: true,
  },
};

const uploadId = await Upload.startUpload(options);
Upload.addListener('progress', uploadId, data => {
  console.log(`[${uploadId}] Progress: ${data.progress}%`);
});
Upload.addListener('error', uploadId, async data => {
  console.log(`[${uploadId}] Error: ${data.error}%`);
});
Upload.addListener('completed', uploadId, async data => {
  console.log(`[${uploadId}] Complete`);
});
```

**Comment by ahansson89** [comment]
Yea. I have ended up writing custom CloudFormation for everything new I am doing in my project. Amplify is just not cutting it.

I wanted to see if I could get around this without creating a signed URL over an API in order to limit the blocking of the UI as much as possible but may end up doing it similar to your solution.

**Comment by osama-rizk** [comment]
Hey everyone, 

 This feature has been released in `aws-amplify@6.16.4`
  (https://github.com/aws-amplify/amplify-js/releases/tag/aws-amplify%406.16.4) (via @aws-amplify/storage@6.14.0).
  You can now generate presigned URLs for uploads using the getUrl API with method: 'PUT':

```
  import { getUrl } from 'aws-amplify/storage';

  const { url } = await getUrl({
    path: 'uploads/file.pdf',
    options: {
      method: 'PUT',
      contentType: 'application/pdf',
      expiresIn: 3600,
    }
  });

  // Use with any HTTP client or custom library
  await fetch(url, {
    method: 'PUT',
    body: file,
    headers: { 'Content-Type': 'application/pdf' },
  });

```

  To upgrade: npm install aws-amplify@latest

Here is the docs link https://docs.amplify.aws/vue/frontend/storage/upload-files/#upload-using-a-presigned-url-1
